ocaml/typing/typecore.ml

```diff
@@ -10004,25 +10006,34 @@ let report_error ~loc env = function
       Location.errorf ~loc ~sub
         "@[This application is complete, but surplus arguments were provided afterwards.@ \
          When passing or calling a local value, extra arguments are passed in a separate application.@]"
-  | Param_mode_mismatch (ty, (_, mkind)) ->
-      let mkind =
+  | Param_mode_mismatch (s, mkind) ->
+      let error_of_equate f (step, {Solver.left; Solver.right}) =
+        let left, right =
+          match (step : equate_step) with
+          | Left_le_right -> left, right
+          | Right_le_left -> right, left
+        in
+        (fun ppf -> f ppf left), (fun ppf -> f ppf right)
+      in
+      let left, right =
         match mkind with
-        | `Locality _ -> "local"
-        | `Uniqueness _ -> "unique"
-        | `Linearity _ -> "once"
+        | `Locality e -> error_of_equate Locality.Const.print (s, e)
+        | `Uniqueness e -> error_of_equate Uniqueness.Const.print (s, e)
+        | `Linearity e -> error_of_equate Linearity.Const.print (s, e)
       in
       Location.errorf ~loc
-        "@[This function has a %s parameter, but was expected to have type:@ %a@]"
-        mkind Printtyp.type_expr ty
+        "@[This function takes a %t parameter,@ \
+         but was expected to take a %t parameter.@]"
+        left right
```

Comment:
Suggestion: run this code in `error_of_equate` (perhaps also renaming `error_of_equate`) so you don't need to capture `left`/`right` in closures just to hide their type.